notebooks/text-generation/mistral-inference-tgi-sagemaker.ipynb

```diff
@@ -0,0 +1,1041 @@
+    "    parser.add_argument(\"--model_dir\", type=str, default=os.environ[\"SM_MODEL_DIR\"])        \n",
+    "    args, _ = parser.parse_known_args()\n",
+    "    \n",
+    "    # parse compilation params\n",
+    "    compilation_params = json.loads(args.compilation_params)\n",
+    "    print(compilation_params)\n",
+    "    # compile and save the model\n",
+    "    model = NeuronModelForCausalLM.from_pretrained(args.model_id, export=True, **compilation_params)\n",
+    "    model.save_pretrained(args.model_dir)\n",
+    "    # now load and export tokenizer\n",
+    "    tokenizer = AutoTokenizer.from_pretrained(args.model_id)\n",
+    "    tokenizer.save_pretrained(args.model_dir)\n",
+    "    print(\"Done!\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3c486648-eecf-4355-951b-bf77c602ac48",
+   "metadata": {},
+   "source": [
+    "#### SageMaker Estimator to kick-off model compilation\n",
+    "If you want to deploy your own model, just change the parameter **--model_id** and point it to the correct HuggingFace repo."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4f6238f0-ac95-47f7-8edb-13d9e4f41fb8",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "import logging\n",
+    "from sagemaker.pytorch import PyTorch\n",
+    "num_cores=2\n",
+    "seq_len=2048\n",
+    "compilation_params= {\"auto_cast_type\": \"bf16\", \"batch_size\": 1, \"sequence_length\": seq_len, \"num_cores\": num_cores}\n",
+    "model_id='yam-peleg/Experiment26-7B'\n",
+    "\n",
+    "estimator = PyTorch(\n",
+    "    entry_point=\"compile.py\", # Specify your train script\n",
+    "    source_dir=\"src\",\n",
+    "    role=role,\n",
+    "    sagemaker_session=sess,\n",
+    "    container_log_level=logging.DEBUG,\n",
+    "    instance_count=1,\n",
+    "    instance_type='ml.trn1.32xlarge',\n",
+    "    output_path=f\"s3://{bucket}/output\",\n",
+    "    disable_profiler=True,\n",
+    "    disable_output_compression=True,\n",
+    "    \n",
+    "    image_uri=f\"763104351884.dkr.ecr.{region}.amazonaws.com/pytorch-training-neuronx:1.13.1-neuronx-py310-sdk2.17.0-ubuntu20.04\",\n",
+    "    \n",
+    "    volume_size = 512,\n",
+    "    hyperparameters={         \n",
+    "        \"compilation_params\": f\"'{json.dumps(compilation_params)}'\",\n",
+    "        \"model_id\": model_id\n",
+    "    }\n",
+    ")\n",
+    "estimator.framework_version = '1.13.1' # workround when using image_uri"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4a0be382-182d-496c-aa10-b690f1a37238",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "estimator.fit()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "585145f3-635a-4bea-b5d8-e6029133acc6",
+   "metadata": {},
+   "source": [
+    "## 2. Deploy the compiled model to SageMaker Endpoint + Inferentia2 + TGI + HF ON"
```

Comment:
You could also have directly exported the model during deployment (saving the first step). Unless I am mistaken the model you chose has the same architecture as zephyr-7b-beta, and is therefore in cache if you use a sequence length of 4096 instead of 2048.

Reply:
I decided to compile it in a separated step due to 2 reasons: 1/ reduce deployment time and avoid timeouts on SageMaker. 2/ When the customer has a custom model they can compile it once and quickly deploy/redeploy multiple times without re-compiling.